### PR TITLE
feat(agw): dockerize the agw envoy controller

### LINKED ISSUE #13684: Fully test containerized AGW
In #13659, a setup to run integration tests of the AGW container deployment was documented. Now this should be extended for more integration tests.

Acceptance criteria:
* [ ] Run all integration tests and try to get them green - document setup, results and possible issues
  * [x] precommit
    * [x] traffic server
  * [x] extended
  * [ ] nonsanity
* [x] Make s1ap_util.py docker aware
  * Look for systemd calls in s1ap_util.py and add a switch for the docker case.
* [x] Set init_system to docker in config for magmad running in the containerzed agw (does not seem to be aleady done (!))
  * Look into running the agw_install_docker.sh script - or set up another way to change the config. @jheidbrink might know more.
  * There is an `init_system` implemented in `magmad/service_manager.py` which sends the command to control docker container. The `docker start` commands are however executed within the magmad-container and throws a `FileNotFoundError: [Errno 2] No such file or directory: 'docker'`
* [x] Enabled magmad to control the hosts docker containers from within the vm (does not seem to be already done (!))
  * Check if this already works if you set the init_system correctly. The magmad container might need access to the docker socket so it can control docker containers outside it.
* [x] Figure out why the envoy controller is not dockerized
  * Probably because it's written in Go. Can it use the C++ container?
  * It might also be because of special dependencies of envoy proxy / controller on the OS side.
  * Relevant integration tests: 
    * s1aptests/test_attach_detach_with_he_policy.py
    * s1aptests/test_attach_detach_rar_tcp_he.py
  * Make sure the systemd envoy controller is not running when executing the tests. It should run as a container.
  *  :warning: From the `journalctl` output after the `docker-compose up`: `Error Loading feg::envoy_controller configs from '/etc/magma/feg/envoy_controller.yml': open /etc/magma/feg/envoy_controller.yml: no such file or directory` 

* [ ] Prepare internal review meeting.

### PR DESCRIPTION
**Pull request by wolfseb**
## Summary

Containerizes the `envoy_controller` in the AGW as a first step for #13684 towards running all integration tests against the dockerized AGW.

* Added envoy controller container to the `docker-compose.yaml` using the go container from `/feg/gateway/docker/go/Dockerfile`
* moved the envoy namespace creation from the `envoy_controller` to the `magma_dp@envoy`, since there is no python env inside the container
* additional feature: added the package `isc-dhcp-client` to be installed in the python container for the pipelined container to work and adapted the `README.md`

## Test Plan

* follow the instructions in the `README.md` to build all the containers
* check that `test_attach_detach` is green

## Additional Information

Created in pairing with @mpfirrmann and @crasu 

- [ ] This change is backwards-breaking

### CODE REVIEW

**Review by Neudrino — DISMISSED** [review]
> * moved the envoy namespace creation from the envoy_controller to the magma_dp@envoy, since there is no python env inside the container

Does the envoy-controller has anything to do with the domain proxy?
In case they are not related at all, this seems very wrong to me, to mix up to clearly distinct domains here, when the usual policy is to have them separated as much as possible.

I would believe, that the DP is an non essential component, which does not need to be deployed. So moving this might be dangerous?

> * Added envoy controller container to the docker-compose.yaml using the go container from /feg/gateway/docker/go/Dockerfile

Same as above: Does the envoy-controller has anything to do with the Federation Gateway?
If not, I would be reluctant to introduce a cross domain dependency here. Those cross domain dependencies make the codebase really hard to maintain, as they are hard to identify and keep track of.
If it also works with the Dockerfile, which was removed, I would see this as preferred solution. Even if there is a coupling than, it would only be loose.

**Comment by mpfirrmann** [comment]
Some quick comments:

- The `magma_dp@envoy` is not related to the domain proxy, but is in fact called "envoy dataplane" ([see here](https://docs.magmacore.org/docs/lte/architecture_overview#header-enrichment)). The two services are closely interconnected.
- The deleted `Dockerfile` was not working at all.

**Comment by mpfirrmann** [comment]
Some quick comments:

> 
> Does the envoy-controller has anything to do with the domain proxy? In case they are not related at all, this seems very wrong to me, to mix up to clearly distinct domains here, when the usual policy is to have them separated as much as possible.
> 
> I would believe, that the DP is an non essential component, which does not need to be deployed. So moving this might be dangerous?
> 
The `magma_dp@envoy` is not related to the domain proxy, but is in fact called "envoy dataplane" ([see here](https://docs.magmacore.org/docs/lte/architecture_overview#header-enrichment)). The two services are closely interconnected.

> 
>  If it also works with the Dockerfile, which was removed, I would see this as preferred solution. Even if there is a coupling than, it would only be loose.


The deleted `Dockerfile` was not working at all.

**Comment by wolfseb** [comment]
> Same as above: Does the envoy-controller has anything to do with the Federation Gateway?

Yes, even in the systemd version installed by `make run`, the envoy-controller is installed from the FeG:
```
build_envoy_controller: ## Build envoy controller
	cd $(MAGMA_ROOT)/feg/gateway && $(MAKE) install_envoy_controller
```

**Review by sebathomas — COMMENTED** [review]
The changes look good to me.

I think the relevant integration tests for the envoy controller are:
-    s1aptests/test_attach_detach_with_he_policy.py
-    s1aptests/test_attach_detach_rar_tcp_he.py

Did you try them too?

**Comment by mpfirrmann** [comment]
> 
> Did you try them too?

Yes. Together with the follow-up PR #13853 we ran them green on both, docker and systemd init systems.

**Review by sebathomas — APPROVED** [review]

File: lte/gateway/docker/docker-compose.yaml [hunk 1]
```diff
@@ -19,6 +19,8 @@ x-standard-volumes: &volumes_anchor
   - /tmp:/tmp
   - /var/log:/var/log
   - /etc/openvswitch:/etc/openvswitch
+  - /home/vagrant/magma/lte/gateway/configs:/magma/lte/gateway/configs
```
Comment (jheidbrink) [inline lte/gateway/docker/docker-compose.yaml]: Shouldn't this be covered by `${CONFIGS_OVERRIDE_VOLUME}:/var/opt/magma/configs`?

File: lte/gateway/docker/docker-compose.yaml [hunk 2]
```diff
@@ -19,6 +19,8 @@ x-standard-volumes: &volumes_anchor
...
+  - /sbin/dhclient:/sbin/dhclient
```
Comment (jheidbrink) [inline lte/gateway/docker/docker-compose.yaml]: I think we should include `dhclient` in the services' docker images if it's needed, not mount it from the host.

File: lte/gateway/docker/docker-compose.yaml (hunk 1 above)
Comment (wolfseb) [reply lte/gateway/docker/docker-compose.yaml]: The configs changed for the first nonsanity test did not come through into the container without mounting these configs. However, this mount explicitly requires the containers to be run inside the vagrant machine. I will look for a better solution.

Comment (jheidbrink) [reply lte/gateway/docker/docker-compose.yaml]: Would it be an option to copy the changed configs to `${CONFIGS_OVERRIDE_VOLUME}` (usually /var/opt/magma/configs)?

Comment (jheidbrink) [reply lte/gateway/docker/docker-compose.yaml]: As discussed with @crasu services probably shouldn't read from `/magma/lte/gateway/configs` anyway. If they insist to do so and it's hard to change, we might keep it for now but then we should create a ticket to improve that.

File: lte/gateway/docker/README.md [hunk 3]
```diff
@@ -41,6 +41,7 @@ containerized AGW by running the following steps inside the VM:
 cd $MAGMA_ROOT/lte/gateway && make run  # You can skip this if you have built the AGW with make before
 for component in redis nghttpx td-agent-bit; do cp "${MAGMA_ROOT}"/{orc8r,lte}/gateway/configs/templates/${component}.conf.template; done
 sudo systemctl stop 'magma@*' 'sctpd' # We don't want the systemd-based AGW to run when we start the containerized AGW
+sudo systemctl start magma_dp@envoy
```
Comment (Neudrino) [inline lte/gateway/docker/README.md:44]: Why is this necessary, when we stop all other services and actually want to have it running as dockerized service and not as systemd service?

File: lte/gateway/docker/docker-compose.yaml [hunk 4]
```diff
@@ -36,6 +38,11 @@ x-lte-cservice: &ltecservice
   <<: *service
   image: ${DOCKER_REGISTRY}agw_gateway_c${OPTIONAL_ARCH_POSTFIX}:${IMAGE_VERSION}
 
+# Generic anchor for go services
+x-goservice: &goservice
```
Comment (Neudrino) [inline lte/gateway/docker/docker-compose.yaml]: I am aware, that the there is inconsistency already above. However, `x-go-service` seems much more readable to me.

_Optional (as out of scope):_ You might also want to refactor to `x-lte-c-service`, in case you value consistency.

File: lte/gateway/docker/docker-compose.yaml [hunk 5]
```diff
@@ -302,3 +309,12 @@ services:
...
     command: /usr/local/bin/liagentd
     restart: "no"
+
+  envoy_controller:
+    <<: *goservice
+    container_name: envoy_controller
+    healthcheck:
+      test: ["CMD", "nc", "-zv", "localhost", "50081"]
+      timeout: "4s"
+      retries: 3
```
Comment (Neudrino) [inline lte/gateway/docker/docker-compose.yaml:317]: Depending on https://github.com/magma/magma/pull/13918, there will be conflicts here. Maybe connect offline.

File: lte/gateway/docker/docker-compose.yaml (hunk 1 above)
Comment (wolfseb) [reply lte/gateway/docker/docker-compose.yaml]: After some testing it is clear that the services actually don't read the configs from `/magma/lte/gateway/configs`, but rather from `/etc/magma` (which are mounted as part of the `$(CONFIGS_DEFAULT_VOLUME)`). I removed the additional mount, because it is unnecessary and unused.

For reference: a test that touches this file to see if everything works is the nonsanity test `test_attach_detach_non_nat_dp_ul_tcp` (remember to run `test_modify_config_for_non_sanity` beforehand)

File: lte/gateway/docker/docker-compose.yaml (hunk 2 above)
Comment (wolfseb) [reply lte/gateway/docker/docker-compose.yaml]: Done.

File: lte/gateway/docker/docker-compose.yaml (hunk 4 above)
Comment (wolfseb) [reply lte/gateway/docker/docker-compose.yaml]: I changed it to a consistent naming scheme: `x-agw-*-service` (including the anchors)

File: lte/gateway/docker/README.md (hunk 3 above)
Comment (jheidbrink) [reply lte/gateway/docker/README.md:44]: This PR dockerizes the envoy controller, not envoy itself (see the diffs in the two .service files above)

File: lte/gateway/docker/services/python/Dockerfile [hunk 6]
```diff
@@ -88,23 +88,24 @@ ARG DEBIAN_FRONTEND=noninteractive
...
+  python3-venv \
   redis-server \
-  wget \
-  ethtool \
-  linux-headers-generic \
-  iptables \
-  iproute2
+  sudo \
+  virtualenv \
+  wget
```
Comment (jheidbrink) [inline lte/gateway/docker/services/python/Dockerfile]: When sorting the list of packages it is not easy as a reviewer to spot changes in this list. I think it would be good to add a comment to this review explaining what has actually changed. This would make it easier for reviewers. Personally I'm a fan of not even mixing functional changes with cosmetic changes for reasons of review readability, but I guess opinions diverge on this topic ;)

Comment (jheidbrink) [reply lte/gateway/docker/services/python/Dockerfile]: Summary of this diff: The package `isc-dhcp-client` has been added.

Comment (wolfseb) [reply lte/gateway/docker/services/python/Dockerfile]: Good point, I reverted the sorting to make it easier to see what happens

**Review by jheidbrink — APPROVED** [review]

File: lte/gateway/docker/services/python/Dockerfile (hunk 6 above)
Comment (Neudrino) [reply lte/gateway/docker/services/python/Dockerfile]: Separation of functional and cosmetic changes is a good idea. But I do not see any issue with both changes in one PR, if one separates those changes out in different commits. Accompanied by a respective comment in the PR description, this works well in my experience.

File: lte/gateway/docker/README.md (hunk 3 above)
Comment (Neudrino) [reply lte/gateway/docker/README.md:44]: I see. Then maybe more a questions of understanding:
It looks to me, if all other Magma services are stopped for the dockerized AGW with `sudo systemctl stop 'magma@*' 'sctpd'`.
Is the envoy service than the only Magma service, which still needs to run on the host for the dockerized AGW?

File: lte/gateway/docker/services/python/Dockerfile (hunk 6 above)
Comment (jheidbrink) [reply lte/gateway/docker/services/python/Dockerfile]: Keep in mind that all commits of a PR will be squashed when merging, so git history will be more difficult to read even when splitting into different commits in the same PR.

**Review by Neudrino — APPROVED** [review]

File: lte/gateway/docker/services/python/Dockerfile (hunk 6 above)
Comment (Neudrino) [reply lte/gateway/docker/services/python/Dockerfile]: - Looking at [a merge commit](https://github.com/magma/magma/commit/15c16eb0bd358553af909a2201f0b0b977724261), I see a reference to [the respective PR](https://github.com/magma/magma/pull/13804), where I can easily access [the single commits](https://github.com/magma/magma/pull/13804/commits/da02598bda0387ef97df335f2f1224064b221013). So I do not see an issue here at all.
- The frequency and amount hardship caused by a chaos state in the current code is much larger, than that of a combined merge commit for functional and cosmetic changes in the past.
